p2p/message.go

```diff
@@ -88,85 +95,47 @@ func (sm *StatusMessage) String() string {
 		sm.ChainStatus)
 }
 
-// Decodes the buffer underlying the reader into a StatusMessage
-// it reads up to specified length
-func (sm *StatusMessage) Decode(r io.Reader, length uint64) (err error) {
-	sm.ProtocolVersion, err = readUint32(r)
-	if err != nil {
-		return err
-	}
-
-	sm.MinSupportedVersion, err = readUint32(r)
-	if err != nil {
-		return err
-	}
-
-	sm.Roles, err = readByte(r)
-	if err != nil {
-		return err
-	}
-
-	sm.BestBlockNumber, err = readUint64(r)
-	if err != nil {
-		return err
-	}
-
-	sm.BestBlockHash, err = readHash(r)
-	if err != nil {
-		return err
-	}
-
-	sm.GenesisHash, err = readHash(r)
+// Encode encodes a status message using SCALE and appends the type byte to the start
+func (sm *StatusMessage) Encode() ([]byte, error) {
+	enc, err := scale.Encode(sm)
 	if err != nil {
-		return err
+		return enc, err
 	}
+	return append([]byte{StatusMsg}, enc...), nil
+}
 
-	if length < 81 {
-		return nil
-	}
-
-	buf := make([]byte, length-81)
-	_, err = r.Read(buf)
-	if err != nil {
-		return err
-	}
-	sm.ChainStatus = buf
-
-	return nil
+// Decodes the message into a StatusMessage, it assumes the type byte has been removed
+func (sm *StatusMessage) Decode(msg []byte) error {
+	_, err := scale.Decode(msg, sm)
+	return err
 }
 
-func readByte(r io.Reader) (byte, error) {
-	buf := make([]byte, 1)
-	_, err := r.Read(buf)
-	if err != nil {
-		return 0, err
-	}
-	return buf[0], nil
+type BlockRequestMessage struct {
+	Id            uint32
+	RequestedData byte
+	StartingBlock []byte      // first byte 0 = block hash (32 byte), first byte 1 = block number (int64)
+	EndBlockHash  common.Hash // optional
+	Direction     byte
+	Max           uint32 // optional
 }
 
-func readUint32(r io.Reader) (uint32, error) {
-	buf := make([]byte, 4)
-	_, err := r.Read(buf)
-	if err != nil {
-		return 0, err
-	}
-	return binary.LittleEndian.Uint32(buf), nil
+// String formats a BlockRequestMessage as a string
+func (bm *BlockRequestMessage) String() string {
+	return fmt.Sprintf("Id=%d RequestedData=%d StartingBlock=0x%x EndBlockHash=0x%x Direction=%d Max=%d",
+		bm.Id,
+		bm.RequestedData,
+		bm.StartingBlock,
+		bm.EndBlockHash,
+		bm.Direction,
+		bm.Max)
 }
 
-func readUint64(r io.Reader) (uint64, error) {
-	buf := make([]byte, 8)
-	_, err := r.Read(buf)
-	if err != nil {
-		return 0, err
-	}
-	return binary.LittleEndian.Uint64(buf), nil
+// Encode encodes a block request message using SCALE and appends the type byte to the start
+func (bm *BlockRequestMessage) Encode() ([]byte, error) {
+	return nil, nil
```

Comment:
Any reason this is included but not fully implemented? Seems you have all the pieces at this point?

Reply:
i wanted to do it with scale but the BlockRequest type has variable/optional types which aren't implemented into scale yet, I do have code for manual encoding I can add though

Reply:
I've added the function and some tests, let me know what you think